crowbar_framework/app/models/service_object.rb

```diff
@@ -1421,6 +1432,25 @@ def proposal_dependencies(role)
     []
   end
 
+  def expand_items_in_elements(elements)
+    # expand items in elements that are not nodes
+    expanded_new_elements = {}
+    failures = []
+    elements.each do |role_name, nodes|
+      expanded_new_elements[role_name], failures = expand_nodes_for_all(nodes)
+      unless failures.nil? || failures.empty?
+        Rails.logger.fatal(
+          "apply_role: Failed to expand items #{failures.inspect} for role \"#{role_name}\""
+        )
+        message = "Failed to apply the proposal: cannot expand list of nodes for role \"#{role_name}\", following items do not exist: #{failures.join(", ")}"
+        Rails.logger.progress("Failed to apply role #{role.name}")
+        update_proposal_status(inst, "failed", message)
```

Comment:
You do not have access to `inst` here; and I probably wouldn't do it here anyway, but in the caller (that's debatable, of course, but it's too magic otherwise). This may also be true for the log bits -- they may be better in the caller.

Reply:
done!